### DISCUSSION TITLE
Variables of this type do not support indexing using points.

### DISCUSSION THREAD

**Discussion by Galaxy2Magic** [yalmip/YALMIP#1057] [Problems using YALMIP] [answered]
出错 Without_sop (line 117)
sol=optimize(C,objective,ops);

`clear all; clc;
warning off;
tic
%% 1.初始数据
%基本参数
mpc=IEEE53;
Pload=mpc.load(:,1);  %有功负荷 KVA
Qload=mpc.load(:,2);  %无功负荷 Kvar
branch=mpc.branch;
r=branch(:,3);  %线路电阻 Ω
x=branch(:,4);  %线路电抗 S
Poz = [2.72;3.2;3.8];  %变压器空载损耗 KW
Pkz = [19.35;20.7;24.3]; 
Sei = [2000;2500;3150];  %变压器容量 KVA
Ssop=3000;  %SOP容量 KVA
nb=53;
nl=52;
Vmax=[1.1*1.1*ones(26,1);1.1;1.1*1.1*ones(15,1);1.1;1.1*1.1*ones(9,1);1.1];  %节点27、43、53电压固定
Vmin=[0.9*0.9*ones(26,1);1.1;0.9*0.9*ones(15,1);1.1;0.9*0.9*ones(9,1);1.1];

upstream=zeros(nb,nl);
dnstream=zeros(nb,nl);
for i=1:nl
    upstream(i,i)=1;
end
for i=[1 3 4 5 8 10 12 13 15 16 18 19 22:26 30 31 33 36 38:42 44 46 49]
    dnstream(i,i+1)=1;
end
    dnstream(1,3)=1;
    dnstream(3,8)=1;
    dnstream(5,7)=1;
    dnstream(8,10)=1;
    dnstream(10,12)=1;
    dnstream(11,22)=1;
    dnstream(13,15)=1;
    dnstream(16,18)=1;
    dnstream(19,21)=1;
    
    dnstream(22,38)=1;
    
    dnstream(24,28)=1;
    dnstream(24,29)=1;
    dnstream(25,30)=1;
    dnstream(26,36)=1;
    dnstream(26,33)=1;
    dnstream(33,35)=1;
 
    dnstream(41,49)=1;
    dnstream(42,44)=1;
    dnstream(42,46)=1;
    dnstream(46,48)=1;
    dnstream(49,51)=1;
    dnstream(49,52)=1;
    dnstream(53,1)=1;
    
   
%% 2.设决策变量
P=sdpvar(nl,1,'full'); %节点流出功率
Q=sdpvar(nl,1,'full');
I=sdpvar(nl,1,'full');
V=sdpvar(nb,1,'full');
Pg=sdpvar(nb,1,'full');
Qg=sdpvar(nb,1,'full');
Pbi=[Pg(27);Pg(43);Pg(53)];
Qbi=[Qg(27);Qg(43);Qg(53)];
Psi=[-P(22);P(23);P(38)];
Qsi=[-Q(22);Q(23);Q(38)];

%% 3.约束
C=[];
 %% 潮流约束
 %节点功率平衡约束
Pin = -upstream*P + upstream*(I.*r) + dnstream*P;  %节点注入有功
Qin = -upstream*Q + upstream*(I.*x) + dnstream*Q;  %节点注入无功
C=[C,Pin+Pload-Pg == 0];
C=[C,Qin+Qload-Qg == 0]; 
%欧姆定律约束
C = [C, V(branch(:,2)) == V(branch(:,1)) - 2*r.*P - 2*x.*Q + (r.^2+x.^2).*I];
%二阶锥约束
C = [C, V(branch(:,1)).*I >= P.^2 + Q.^2];

 %% 通用约束
%节点电压约束
C = [C, Vmin <= V,V <= Vmax];
%支路电流约束
C = [C, 0 <= I,I <= 0.11];
%支路功率约束
C = [C, -0.11<=P,P <= 0.11,-0.11 <= Q,Q <= 0.11];

%% 器件约束
%SOP容量约束
% C=[C,Psi(1) == -P(22);Psi(2)==P(23);Psi(3)==P(38)];  %SOP有功功率
% C=[C,Qsi(1) == -Q(22);Qsi(2)==Q(23);Qsi(3)==Q(38)];%SOP无功功率

C=[C,sqrt(Psi.^2+Qsi.^2)<=3000];
%变压器容量约束
%C=[C,Pbi==[Pg(27);Pg(43);Pg(53)];Qbi==[Qg(27);Qg(43);Qg(53)]];
C=[C,sqrt(Pbi.^2+Qbi.^2)<=Sei];
%变压器最佳经济运行区间约束
belta2 =(Pbi.^2+Qbi.^2)./(Sei.^2);
C=[C,0.0625<=belta2,belta2<=0.5625];
%% 4.目标函数
%变压器综合损耗
Pzi=sum(Poz+belta2.*Pkz);
%SOP端口传输损耗
Psi_loss=sum(0.02*sqrt(Psi.^2+Qsi.^2));
%线路损耗
Ploss=sum(I.*r);
%总损耗
objective=Pzi+Psi_loss+Ploss;
toc %建模时间
%% 5.设置求解器求解
ops=sdpsettings('verbose',1,'solver','cplex');
% ops.fmincon.MaxIter=100000;
% ops.fmincon.MaxFunEvals=1000000;
sol=optimize(C,objective,ops);
toc%求解时间
%% 6.分析错误标志
if sol.problem == 0
    disp('succcessful solved');
else
    disp('error');
    yalmiperror(sol.problem)
end
`

**Comment by johanlofberg** [comment]
When you copy code, select it afterwards and put a code block <> on it. As it is now, the display is broken, which you also should see

**Reply by Galaxy2Magic** [reply]
```
%三变压器53节点系统，SOP看为一个节点，系统基准容量维1MVA,基准电压11KV ，利用矩阵构造网络  定义一个Pg,Pbi,Psi
%均用Pg中元素表示，然后再增添约束
clear all; clc;
warning off;
tic
%% 1.初始数据
%基本参数
mpc=IEEE53;
Pload=mpc.load(:,1);  %有功负荷 KVA
Qload=mpc.load(:,2);  %无功负荷 Kvar
branch=mpc.branch;
r=branch(:,3);  %线路电阻 Ω
x=branch(:,4);  %线路电抗 S
Poz = [2.72;3.2;3.8];  %变压器空载损耗 KW
Pkz = [19.35;20.7;24.3]; 
Sei = [2000;2500;3150];  %变压器容量 KVA
Ssop=3000;  %SOP容量 KVA
nb=53;
nl=52;
Vmax=[1.1*1.1*ones(26,1);1.1;1.1*1.1*ones(15,1);1.1;1.1*1.1*ones(9,1);1.1];  %节点27、43、53电压固定
Vmin=[0.9*0.9*ones(26,1);1.1;0.9*0.9*ones(15,1);1.1;0.9*0.9*ones(9,1);1.1];

upstream=zeros(nb,nl);
dnstream=zeros(nb,nl);
for i=1:nl
    upstream(i,i)=1;
end
for i=[1 3 4 5 8 10 12 13 15 16 18 19 22:26 30 31 33 36 38:42 44 46 49]
    dnstream(i,i+1)=1;
end
    dnstream(1,3)=1;
    dnstream(3,8)=1;
    dnstream(5,7)=1;
    dnstream(8,10)=1;
    dnstream(10,12)=1;
    dnstream(11,22)=1;
    dnstream(13,15)=1;
    dnstream(16,18)=1;
    dnstream(19,21)=1;
    
    dnstream(22,38)=1;
    
    dnstream(24,28)=1;
    dnstream(24,29)=1;
    dnstream(25,30)=1;
    dnstream(26,36)=1;
    dnstream(26,33)=1;
    dnstream(33,35)=1;
 
    dnstream(41,49)=1;
    dnstream(42,44)=1;
    dnstream(42,46)=1;
    dnstream(46,48)=1;
    dnstream(49,51)=1;
    dnstream(49,52)=1;
    dnstream(53,1)=1;
    
   
%% 2.设决策变量
P=sdpvar(nl,1,'full'); %节点流出功率
Q=sdpvar(nl,1,'full');
I=sdpvar(nl,1,'full');
V=sdpvar(nb,1,'full');
Pg=sdpvar(nb,1,'full');
Qg=sdpvar(nb,1,'full');
Pbi=[Pg(27);Pg(43);Pg(53)];
Qbi=[Qg(27);Qg(43);Qg(53)];
Psi=[-P(22);P(23);P(38)];
Qsi=[-Q(22);Q(23);Q(38)];

%% 3.约束
C=[];
 %% 潮流约束
 %节点功率平衡约束
Pin = -upstream*P + upstream*(I.*r) + dnstream*P;  %节点注入有功
Qin = -upstream*Q + upstream*(I.*x) + dnstream*Q;  %节点注入无功
C=[C,Pin+Pload-Pg == 0];
C=[C,Qin+Qload-Qg == 0]; 
%欧姆定律约束
C = [C, V(branch(:,2)) == V(branch(:,1)) - 2*r.*P - 2*x.*Q + (r.^2+x.^2).*I];
%二阶锥约束
C = [C, V(branch(:,1)).*I >= P.^2 + Q.^2];

 %% 通用约束
%节点电压约束
C = [C, Vmin <= V,V <= Vmax];
%支路电流约束
C = [C, 0 <= I,I <= 0.11];
%支路功率约束
C = [C, -0.11<=P,P <= 0.11,-0.11 <= Q,Q <= 0.11];

%% 器件约束
%SOP容量约束
% C=[C,Psi(1) == -P(22);Psi(2)==P(23);Psi(3)==P(38)];  %SOP有功功率
% C=[C,Qsi(1) == -Q(22);Qsi(2)==Q(23);Qsi(3)==Q(38)];%SOP无功功率

C=[C,sqrt(Psi.^2+Qsi.^2)<=3000];
%变压器容量约束
%C=[C,Pbi==[Pg(27);Pg(43);Pg(53)];Qbi==[Qg(27);Qg(43);Qg(53)]];
C=[C,sqrt(Pbi.^2+Qbi.^2)<=Sei];
%变压器最佳经济运行区间约束
belta2 =(Pbi.^2+Qbi.^2)./(Sei.^2);
C=[C,0.0625<=belta2,belta2<=0.5625];
%% 4.目标函数
%变压器综合损耗
Pzi=sum(Poz+belta2.*Pkz);
%SOP端口传输损耗
Psi_loss=sum(0.02*sqrt(Psi.^2+Qsi.^2));
%线路损耗
Ploss=sum(I.*r);
%总损耗
objective=Pzi+Psi_loss+Ploss;
toc %建模时间

%% 5.设置求解器求解
ops=sdpsettings('verbose',1,'solver','cplex');
sol=optimize(C,objective,ops);
toc%求解时间
%% 6.分析错误标志
if sol.problem == 0
    disp('succcessful solved');
else
    disp('error');
    yalmiperror(sol.problem)
end

**Reply by Galaxy2Magic** [reply]
For the first time, I'm sorry I didn't notice this problem

**Reply by johanlofberg** [reply]
Unrecognized function or variable 'IEEE53'.
 
>> 

You have to supply all data required / make example independent of any particular data

**Reply by Galaxy2Magic** [reply]
IEEE53.m
bus_22 is a SOP(Soft Open Switch) 
```
function mpc=IEEE53
IEEE22=case22;
mpc.branch=[53 1 0.3664 0.1807;1 2 0.0547 0.0282;1 3 0.5416 0.2789;3 4 0.193 0.099;4 5 1.311 0.6752;5 6 0.0598 0.0308;5 7 0.2905 0.1496;
            3 8 0.7431 0.3827;8 9 0.0547 0.0282;8 10 0.675 0.3481;10 11 0.0547 0.0282;10 12 0.3942 0.203;12 13 1.046 0.5388;13 14 0.022 0.0116;
             13 15 0.0547 0.0282;15 16 0.3212 0.1654;16 17 0.0949 0.0488;16 18 0.547 0.2959; 18 19 0.1292 0.066;19 20 0.0871 0.045;
             19 21 0.5329 0.2744;11 22 0 0;22 23 0 0;23 24 1.52348 1.0276;24 25 0.84111 0.82271;25 26 1.17024 1.14464;26 27 1.35309 1.32349;
             24 28 1.19702 0.8074;24 29 2.23081 1.5047;25 30 1.79553 1.2111;30 31 2.44845 1.6515;31 32 2.01317 1.3579;26 33 2.55727 1.7249;
             33 34 1.25143 0.8441;33 35 1.0882 0.734;26 36 2.01317 1.3579;36 37 1.68671 1.1377;22 38 0 0;38 39 2.01317 1.3579;39 40 2.44845 1.6515;
             10 41 1.79553 1.2111;41 42 1.17024 1.14464;42 43 1.35309 1.32349;42 44 2.01317 1.3579;44 45 1.68671 1.1377;42 46 2.55727 1.7249;
             46 47 1.0882 0.734;46 48 1.25134 0.8441;41 49 0.84111 0.82271;49 50 1.19702 0.8074;49 51 2.23081 1.5047;49 52 1.52348 1.0276];
mpc.load=[16.78 20.91;16.78 20.91;33.8 37.32;14.56 12.52;10.49 14.21;8.821 11.66;14.35 18.59;19.31 25.87;14.35 18.59;16.27 19.48;16.27 19.48;
          82.13 71.65;34.71 30.12;34.71 30.12;80.31 70.12;49.62 47.82;49.62 47.82;43.77 38.93;37.32 35.96;37.32 35.96;31.02 29.36;0 0;44.1 44.991;
          140 142.8286;70 71.4143;44.1 44.991;0 0;140 142.8286;70 71.4143;140 142.8286;70 7.4143;44.1 44.991;140 142.8286;70 71.4143;140 142.8286;
          70 71.4143;44.1 44.991;44.1 44.991;70 71.4143;140 142.8286;70 71.4143;44.1 44.991;0 0;70 71.4143;44.1 44.991;140 142.8286;140 142.8286; 
          70 71.4143;140 142.8286;140 142.8286;70 71.4143;44.1 44.991;0 0];

**Comment by johanlofberg** [comment, marked as answer]
Cannot reproduce your error (which you really haven't explained where and how it shows up)

However, this cannot be solved using cplex

```
sol = 

  struct with fields:

    solvertime: 0
          info: 'Solver not applicable (cplex does not support nonconvex quadratic constraints)'
       problem: -4
    yalmiptime: 0.7300
```

**Reply by Galaxy2Magic** [reply]
![error](https://user-images.githubusercontent.com/92700668/147332559-88a60852-a08d-493d-8bcc-787b798b422f.jpg)

**Reply by johanlofberg** [reply]
Also note that if you switch to gurobi, it will immediately claim it is infeasible, and digging down shows that gurobi e.g. thinks the two last quadratic constraints are infeasible

```
sol=optimize(C(15:16),[],sdpsettings('solver','gurobi'))

    yalmipversion: '20210331'
    matlabversion: '9.9.0.1524771 (R2020b) Update 2'
       yalmiptime: 0.1507
       solvertime: 0.2823
             info: 'Infeasible problem (learn to debug) (GUROBI-NONCONVEX)'
          problem: 1
```


However, switching to YALMIPs built-in global solver bmibnb, a solution is found to those two constraints, hence gurobi is wrong here

The two constraints are effectively something like the following annulus constraints

```
P = [0.0625 <= 1e-7*[(x^2+y^2);
                     (u^2+v^2);
                     (a^2+b^2)] <= .5625];

optimize(P,[],sdpsettings('solver','gurobi'))
```

Trivially feasible, but gurobi claims infeasible. The problem is the bad scaling. With a better scaling of a similar problem, gurobi has no issues. For instance the following is solved easily

```
P = [0.0625 <= 1e-5*[(x^2+y^2);
                     (u^2+v^2);
                     (a^2+b^2)] <= .5625];
optimize(P,[],sdpsettings('solver','gurobi'))
```

**Reply by johanlofberg** [reply]
Looks like installation of YALMIP is broken on your machine (or you are using a very very very old version of either yalmip or matlab). I would start by deleting the current installation, and download the latest version, and make sure it actually works first (by solving a trivial silly problem)

**Reply by Galaxy2Magic** [reply]
Thank you for your suggestions! I will install the latest version and run my program again.If there is some questions, I'll consult you again.
